### PR TITLE
[FLINK-18646] Verify memory manager empty in a separate thread with l…

### PR DESCRIPTION
**Pull request by azagrebin**
`UnsafeMemoryBudget#verifyEmpty`, called on slot freeing, needs time to wait on GC of all allocated/released managed memory. If there are a lot of segments to GC then it can take time to finish the check. If slot freeing happens in RPC thread, the GC waiting can block it and TM risks to miss its heartbeat.

Another problem is that after `UnsafeMemoryBudget#RETRIGGER_GC_AFTER_SLEEPS`, `System.gc()` is called for each attempt to run a cleaner even if there are already detected cleaners to run. This leads to triggering a lot of unnecessary GCs in background.

The PR offloads the verification into a separate thread and calls `System.gc()` only if memory cannot be reserved and there are still no cleaners to run after long waiting. The timeout for normal memory reservation is increased to 2 second. The full reservation, used for verification, gets 2 minute timeout.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/memory/UnsafeMemoryBudget.java [hunk 1]
```diff
@@ -37,9 +37,10 @@
 class UnsafeMemoryBudget {
 	// max. number of sleeps during try-reserving with exponentially
 	// increasing delay before throwing OutOfMemoryError:
-	// 1, 2, 4, 8, 16, 32, 64, 128, 256, 512 (total 1023 ms ~ 1 s)
+	// 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, ... (total (2 x 1024) - 1 ms ~ 2 s)
 	// which means that MemoryReservationException will be thrown after 1 s of trying
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/memory/UnsafeMemoryBudget.java:41]:
```suggestion
	// which means that MemoryReservationException will be thrown after 2 s of trying
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java [hunk 2]
```diff
@@ -84,15 +85,22 @@
 	/** The closing future is completed when the slot is freed and closed. */
 	private final CompletableFuture<Void> closingFuture;
 
+	/**
+	 * {@link ExecutorService} for background actions, e.g. verify all managed memory released.
+	 */
+	private final ExecutorService asyncExecutor;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java:91]:
```suggestion
	private final Executor asyncExecutor;
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java [hunk 3]
```diff
@@ -84,15 +85,22 @@
 	/** The closing future is completed when the slot is freed and closed. */
 	private final CompletableFuture<Void> closingFuture;
 
+	/**
+	 * {@link ExecutorService} for background actions, e.g. verify all managed memory released.
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java:89]:
```suggestion
	 * {@link Executor} for background actions, e.g. verify all managed memory released.
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 4]
```diff
@@ -134,6 +143,12 @@ public TaskSlotTableImpl(
 		slotActions = null;
 		state = State.CREATED;
 		closingFuture = new CompletableFuture<>();
+
+		asyncExecutor = new ThreadPoolExecutor(
+			0,
+			numberSlots,
+			60L, TimeUnit.SECONDS,
+			new SynchronousQueue<>());
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:151]:
```suggestion
			new SynchronousQueue<>(),
			new ExecutorThreadFactory("task-slot-memory-verifier"));
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 5]
```diff
@@ -107,6 +111,11 @@
 		"TaskSlotTableImpl is not initialized with proper main thread executor, " +
 			"call to TaskSlotTableImpl#start is required");
 
+	/**
+	 * {@link ExecutorService} for background actions, e.g. verify all managed memory released.
+	 */
+	private final ExecutorService asyncExecutor;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:117]:
```suggestion
	private final ExecutorService memoryVerificationExecutor;
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java [hunk 6]
```diff
@@ -134,6 +143,12 @@ public TaskSlotTableImpl(
 		slotActions = null;
 		state = State.CREATED;
 		closingFuture = new CompletableFuture<>();
+
+		asyncExecutor = new ThreadPoolExecutor(
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:147]: What spoke against using the `ioExecutor` in `TaskManagerSharedServices` when creating the `TaskSlotTableImpl`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java [hunk 7]
```diff
@@ -295,22 +303,25 @@ public String toString() {
...
+
 			final CompletableFuture<Void> cleanupFuture = FutureUtils
 				.waitForAll(tasks.values().stream().map(TaskSlotPayload::getTerminationFuture).collect(Collectors.toList()))
-				.thenRun(() -> {
-					verifyMemoryFreed();
-					this.memoryManager.shutdown();
-				});
-
+				.thenRun(memoryManager::shutdown);
+			verifyAllManagedMemoryIsReleasedAfter(cleanupFuture);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java:310]: Maybe add a test which ensures that we can call `MemoryManager.verifyEmpty()` after it has been shut down. That way we explicitly state that this an explicit contract we want to guard.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java [hunk 8]
```diff
@@ -295,22 +303,25 @@ public String toString() {
...
-		}
+	private void verifyAllManagedMemoryIsReleasedAfter(CompletableFuture<Void> after) {
+		FutureUtils.runAfterwardsAsync(
+			after,
+			() -> {
+				if (!memoryManager.verifyEmpty()) {
+					LOG.warn("Not all slot memory is freed, potential memory leak at {}", this);
+				}
+			},
+			asyncExecutor);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java:324]: Does it make sense to only run this action if `after` has completed normally. E.g. using `after.thenRunAsync(() -> ..., asyncExecutor)`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java:91]:
```suggestion
	private final ExecutorService memoryVerificationExecutor;
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java (hunk 6 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImpl.java:147]: I think we can indeed use the `ioExecutor`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java (hunk 8 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlot.java:324]: yes, I think so

**Comment by azagrebin** [comment]
Thanks for the review @tillrohrmann 
I addressed the comment and some of @zentol's from #12981

**Pull request closed without merging**